src/Operation/BulkWrite.php

```diff
@@ -330,13 +341,33 @@ public function execute(Server $server)
                     break;
 
                 case self::INSERT_ONE:
-                    $insertedIds[$i] = $bulk->insert($args[0]);
+                    $insertedDocument = isset($this->options['codec'])
+                        ? $this->options['codec']->encodeIfSupported($args[0])
+                        : $args[0];
+                    // Psalm's assert-if-true annotation does not work with unions, so
+                    // assert the type manually instead of using is_document
+                    // See https://github.com/vimeo/psalm/issues/6831
+                    assert(is_array($insertedDocument) || is_object($insertedDocument));
```

Comment:
Similar to my suggestion in FindOneAndReplace, I think this assertion is only relevant for `if (isset($this->options['codec']))`. The constructor already checks `is_document($args[0])` for all code paths, and `is_document($args[1])` for replaceOne.

In this case, I don't object to something like:

```php
if (isset($this->options['codec'])) {
    $args[0] = $this->options['codec']->encodeIfSupported($args[0]);
    // Psalm comment...
    assert(is_array($args[0]) || is_object($args[0]));
}
```

And likewise for replaceOne below. Doing so would also allow us to keep the fall through to a common `$bulk->update($args[0], $args[1], $args[2]);`. I'd suggest adding a `// Fall through` comment after the `REPLACE_ONE` case like we do in the test suite in various classes that deal with spec test operations.

Reply:
Updated along with a comment documenting that the missing `break` is intentional.